File: examples/gameroom/gameroom-app/src/utils/addressing.ts
// Copyright 2019 Cargill Incorporated
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const crypto = require('crypto');

export const XO_FAMILY_NAME = 'xo';
export const XO_FAMILY_VERSION = '0.3.3';
export const XO_FAMILY_PREFIX = '5b7349';

// The namespace registry prefix for global state (00ec00)
const NAMESPACE_REGISTRY_PREFIX = '00ec00';

// The contract prefix for global state (00ec02)
const CONTRACT_PREFIX = '00ec02';

// The contract registry prefix for global state (00ec01)
const CONTRACT_REGISTRY_PREFIX = '00ec01';

export function calculateGameAddress(gameName: string): string {
  const gameNameHash =  crypto.createHash('sha512').update(gameName).digest('hex');
  return `${XO_FAMILY_PREFIX}${gameNameHash.slice(0, 64)}`;

}

export function calculateNamespaceRegistryAddress(namespace: string) {
    const prefix = namespace.slice(0, 6);
    const hash = crypto.createHash('sha512').update(prefix).digest('hex').slice(0, 64);
    return `${NAMESPACE_REGISTRY_PREFIX}${hash}`;
}


export function computeContractAddress(name: string, version: string) {
    const input = `${name},${version}`;
    const hash = crypto.createHash('sha512').update(input).digest('hex').slice(0, 64);
    return `${CONTRACT_PREFIX}${hash}`;
}

export function computeContractRegistryAddress(name: string) {
    const hash = crypto.createHash('sha512').update(name).digest('hex').slice(0, 64);
    return `${CONTRACT_REGISTRY_PREFIX}${hash}`;
}

File: sdk/src/product/addressing.rs
// Copyright 2018-2021 Cargill Incorporated
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

pub const GRID_NAMESPACE: &str = "621dee";
pub const PRODUCT_PREFIX: &str = "02";
pub const GRID_PRODUCT_NAMESPACE: &str = "621dee02";

/// Computes the address of a GS1 product based on its GTIN
pub fn compute_gs1_product_address(gtin: &str) -> String {
    // 621ddee (grid namespace) + 02 (product namespace) + 01 (gs1 namespace)
    String::from(GRID_NAMESPACE)
        + PRODUCT_PREFIX
        + "01"
        + "00000000000000000000000000000000000000000000"
        + &format!("{:0>14}", gtin)
        + "00"
}
